### ISSUE TITLE
Evaluate in scope

### ISSUE THREAD

**Issue by caspervonb** [caspervonb/amok#135]
Split from #79, the problem we are trying to solve is how can we actually evaluate in scope.

The simplest approach would be to just instrument the code with evaluate hooks
```js
var __functions = [];

function __pry(fn) {
  __functions.push(fn);
}

(function() {
  var value = 0;

  __pry(function(code) {
    return eval(code);    
  });

  setInterval(function() {
    console.log('value: ' + value);
  }, 1000);
}());
```

Extremely simple and easy to reason about, since its within the closure we have access to the closure.
This could quite easily be instrumented in a pass via a proxy server, so that the compiler pipeline remains untouched.

But there are caveats to this approach, since the eval is in a callback it has its own local scope, so we can't define local variables in the closure itself.

For module patterns this translates to being unable to do imports and create new private state and helper functions.

**Comment by caspervonb** [comment]
An alternate approach is to go via the debugger, the tip of the tree protocol has a couple of new methods for dealing with scopes.

Namely `Debugger.getFunctionDetails` which returns the scope chain of a function and `Debugger.setVariableValue`.

Now as far as I can tell you still need a reference to a function in order get its `objectId` which is required to call `Debugger.getFunctionDetails`.

I have been able to get the scope chains from functions, from there its closure but have not been able to get `Debugger.setVariableValue` to have any effect so far.

Think I'll have to dig into the source v8 source once again to figure out if this is viable at all.

Both approaches requires instrumentation, for the debugger approach we still have to track the scopes ourselves.

```js
__functions = [];

function __pry(fn) {
  __functions.push(fn);
}

(function __scope_01() {
  var value = 0;

  __pry(__scope_01);

  setInterval(function() {
    console.log('value: ' + value);
  }, 1000);
}());
```

**Comment by stereokai** [comment]
Thanks @caspervonb. I want to help, but I'm not sure I understand the first comment.

**Comment by caspervonb** [comment]
**TL;DR** Loaders/transform solve this to some degree.

Opting to just do this at the compiler level for user modules (#136), will revisit when v8 modules are a thing.

The gist of this, is fairly straight forward, function declarations and variable declarations are stuck into a `locals` object.

Wrapping this in a with statement everything appears like normal, it's a cheap trick but the runtime capabilities of JavaScript is nothing like, for example ruby where all this nonsense would just be a call to `scope.local_variable_set`.

So, we'll be exposing an exec function on the module object through webpack/browserify injection.
```js
function exec(code) {
  var fn = new Function('module', 'exports', 'require', 'locals', [
    'with (locals) {',
    '  return ' + code,
    '}'
  ].join('\n'));

  return fn(module, exports, __webpack_require__, locals);
}

module.exec = exec;
```

The *REPL* can then be given commands to probe for `require.cache` in order to find  our little pseudo execution contexts. So this will basically come nearly free, the above function is an internal function for the re-execution thing I'm working on for #136

**Comment by caspervonb** [comment]
Also dealing with this at the bundler/loader level, might expose a pattern for some quicker access via commands but until then, closed.
